src/databricks/labs/ucx/migration/azure_credentials.py

```diff
@@ -0,0 +1,285 @@
+            name=storage_credential.name,
+            azure_service_principal=storage_credential.azure_service_principal,
+            created_by=storage_credential.created_by,
+            read_only=storage_credential.read_only,
+            results=validation.results,
+        )
+
+
+class AzureServicePrincipalMigration:
+
+    def __init__(
+        self,
+        installation: Installation,
+        ws: WorkspaceClient,
+        azure_resource_permissions: AzureResourcePermissions,
+        azure_sp_crawler: AzureServicePrincipalCrawler,
+    ):
+        self._output_file = "azure_service_principal_migration_result.csv"
+        self._final_sp_list: list[ServicePrincipalMigrationInfo] = []
+        self._installation = installation
+        self._ws = ws
+        self._azure_resource_permissions = azure_resource_permissions
+        self._azure_sp_crawler = azure_sp_crawler
+        self._action_plan = 'service_principals_for_storage_credentials.csv'
+
+    @classmethod
+    def for_cli(cls, ws: WorkspaceClient, product='ucx'):
+        installation = Installation.current(ws, product)
+        config = installation.load(WorkspaceConfig)
+        sql_backend = StatementExecutionBackend(ws, config.warehouse_id)
+        azurerm = AzureResources(ws)
+        locations = ExternalLocations(ws, sql_backend, config.inventory_database)
+
+        azure_resource_permissions = AzureResourcePermissions(installation, ws, azurerm, locations)
+        azure_sp_crawler = AzureServicePrincipalCrawler(ws, sql_backend, config.inventory_database)
+
+        return cls(installation, ws, azure_resource_permissions, azure_sp_crawler)
+
+    def _list_storage_credentials(self) -> set[str]:
+        # list existed storage credentials that is using service principal, capture the service principal's application_id
+        storage_credential_app_ids = set()
+
+        storage_credentials = self._ws.storage_credentials.list(max_results=0)
+        for storage_credential in storage_credentials:
+            # only add service principal's application_id, ignore managed identity based storage_credential
+            if storage_credential.azure_service_principal:
+                storage_credential_app_ids.add(storage_credential.azure_service_principal.application_id)
+        logger.info(
+            f"Found {len(storage_credential_app_ids)} distinct service principals already used in UC storage credentials"
+        )
+        return storage_credential_app_ids
+
+    def _check_sp_in_storage_credentials(self, sp_list, sc_set) -> list[StoragePermissionMapping]:
+        # if sp is already used, take it off from the sp_list
+        filtered_sp_list = []
+        for service_principal in sp_list:
+            if service_principal.client_id not in sc_set:
+                filtered_sp_list.append(service_principal)
+
+        return filtered_sp_list
+
+    def _read_databricks_secret(self, scope: str, key: str, application_id: str) -> str | None:
+        try:
```

Comment:
Why do we need to catch all these exceptions can we recover from any of these. We can just call the get_secret and let the user respond to the exception.

Reply:
Removed the catch for authentication and permission denied, as such exception should be rethrow and fail the command. For the rest InternalError and ResourceDoesNotExist we need to catch it and continue the command execution.